active/0000-bicep-version-pinning.md

```diff
@@ -0,0 +1,253 @@
+- `1.*` - equivalent to `>=1.0.0,<2.0.0`
+- `1.2.*` - equivalent to `>=1.2.0,<1.3.0`
+- `*` -  equivalent to `>=0.0.0`
+
+However, we should impose the following restrictions:
+1. The wilcard `*` cannot be used together with the version comparison syntax. 
+    - Reason: This is to prevent ambiguous snytaxes like this: `>=1.*`. It is hard to reason about the allowed range in this case, i.e. does the range start from `1.0.0` or `1.1.2` or `1.3.5` or something else?
+2. If a version number is replaced with a wildcard, then all later version numbers must also be wildcards. Thus `2.*.6` is an invalid wildcard version. 
+    - Reason: Most semver parsing tools (including the [tools sampled above](#option-2-npmterraform-syntax)) either do not support this, or they ignore anything after the `*`.
+
+### VSCode experience
+Because the VSCode extension version is tightly coupled to the Bicep compiler version. If we fail the Bicep compilation, the user if forced to update their VSCode extension version; this is not a great user experience especially when working across multiple Bicep repos that they even might not own. 
+
+Ideally, we would decouple the extension from the Bicep version, but that would be an expensive undertaking.
+
+An acceptable solution here is: instead of throwing compiler errors, we'd warn the user if their VSCode extension version doesn't match the configured version constraint. 
+This way the user is aware of the mismatch, but we do not completely block them from proceeding with compilation.
+We will need to clearly document this behavior in which the pinning is not strictly enforced in VSCode, so that users do not have the wrong expectations.
+
+
+### AzCLI updates
+#### Basic mechanism
+- The AzCLI bicep module should be enhanced to parse the appropriate `bicepconfig.json` file by recursively searching upwards from the current directory. It should then install the latest Bicep version that satisfies the specified version constraints. 
+    - For example, given the version constraint `>=0.31.0 <0.32.0` and the available Bicep release tags `[0.21.14, 0.31.1, 0.31.92, 0.32.45]`, AzCLI would download version `0.31.92`.
+- If no version constraints are found, AzCLI should default to using locally installed version (or downloading the latest version if no Bicep is not installed locally).
+
+#### Caching
+- In order to minimize unnecessary downloads and optimize bandwidth usage especially for users switching between projects that have version constraints, AzCLI should cache the downloaded version on disk, for example as `~\.azure\bin\bicepVersions\bicep_<version>.exe`. Then when a locally available version is needed, switch the bicep version by copying the binary into `~\.azure\bin\bicep.exe`. The reason for copying the binary is for backward-compatibility because currently AzCLI invokes bicep commands via the binary located in `~\.azure\bin\bicep.exe`.
```

Comment:
It feels like we can do this with symbolic links, but they seem difficult to use in Windows.